### PR TITLE
Default pack/unpack for non packable/unpackable data.

### PR DESCRIPTION
**Pull request by TotoGaz**
This PR solves https://github.com/GEOSX/GEOSX/issues/1848 by...
- ...making the "_wrapper exclusion pattern_" a bit more flexible since it's now allowed to dynamically register wrappers that will not be packed (previous pattern had was more static). This pattern is already used thrice. (and the `"usedFaces"` trick now abides by the new pattern).
- ...eventually setting errors when the default functions are called. This would tell any user that something unintended happened and needs to be investigated: no _wrong_ data will be provided by mistake to the end user.

- [x] Passes the integrated tests (baselines are unchanged).

### CODE REVIEW

**Comment by rrsettgast** [comment]
I think the "do nothing" option should be removed and we should have a logical approach to avoid calling pack/unpack on things that cannot be packed and unpacked. Also, the example of a `geosx::InterObjectRelation<LvArray::Array<long, 2, camp::int_seq<long, 0l, 1l>, long, LvArray::ChaiBuffer> >` is not such a case...we should be packing/unpacking this, so we should just implement a specialization for it that calls the pack/unpack for `LvArray::::Array<long, 2, camp::int_seq<long, 0l, 1l>, long, LvArray::ChaiBuffer>` and sets the pointer...If @corbett5 were to chip in, he would say that we should just do away with the `InterObjectRelation` types. They do circumvent encapsulation, but the pointers to the related objects are used for cases where we need to access `localToGlobal` and `globalToLocal`...or at least they were at one point. If they are no longer used, then we just ditch the relation types.

**Comment by TotoGaz** [comment]
I personally expect the _"do nothing"_ implementations to be some convenience C++ implementations such that _"it compiles without adding complicated templates"_. (It's just fortune telling).

If we want to remove `geosx::InterObjectRelation`, would replacing by the `LvArray` it inherits from enough? I do not have enough information on what we expect from this `class`.

**Comment by rrsettgast** [comment]
Indeed the "do nothing" options were likely put in to avoid having to implement everything. It was fine to get things working, but it is a bit unclean to call pack/unpack functions on something and have it not actually be packed/unpacked.

I suspect that getting rid of the relations will take some work since they should have their own pack/unpack functions.

**Comment by wrtobin** [comment]
I'd agree that the do-nothing should be replaced with an error-out in principle and for code correctness, though the dev time to remove the error from occurring might as Randy says require nontrivial effort.

**Comment by TotoGaz** [comment]
I've run the `integratedTests` on my box (up to 6 ranks, but the errors will be the same).
I could spot one unique failure:
```
Trying to pack data type (geosx::InterObjectRelation<LvArray::Array<long, 2, camp::int_seq<long, 0l, 1l>, long, LvArray::ChaiBuffer> >) by index. Operation not supported.
```

**Comment by TotoGaz** [comment]
I selected the `HydrofracturingSolver/kgdValidation_smoke.xml` integrated test (running on 4 ranks) because it exhibits the three types that are silently ignored by our packing by index functions (which are called nevertheless, see [comment](https://github.com/GEOSX/GEOSX/pull/1788#issuecomment-1050465415)).

I've debugged which wrappers pass through the dummy `packByIndex` implementations.

- "_nodeList_" as a `geosx::InterObjectRelation<LvArray::Array<long, 2, camp::int_seq<long, 0l, 1l>, long, LvArray::ChaiBuffer> >`
- "_fracturedCells_" as a `LvArray::SortedArray<long, long, LvArray::ChaiBuffer>` 
- "_FE1_" as a `geosx::finiteElement::FiniteElementBase`.

Those quantities are "_packed_" multiple times with different packLists. They are "_unpacked_" as well.
Most likely the information is not exchanged.

This surely impacts the result/validity of this integrated test? @castelletto1 @CusiniM

**Comment by TotoGaz** [comment]
Thx for the information @CusiniM 

> So, for these, checking whether they are `sizedFromParent` or not is actually an effective criterion to know whether they should be packed by index or not because we really should not need to pack/unpack them.

Isn't this actually done?

For information, here are the integrated tests that I could run on my box (6 cores) and which go through the dummy implementation (there are duplications because of number of MPI ranks, obviously).
```
HydrofracturingSolver/Sneddon_hydroFrac04
HydrofracturingSolver/heterogeneousInSitu_smoke_04
HydrofracturingSolver/kgdToughnessDominated_smoke_06
HydrofracturingSolver/kgdValidation_smoke_04
HydrofracturingSolver/kgdViscosityDominated_smoke_06
HydrofracturingSolver/pennyShapedToughnessDominated_smoke_06
compositionalMultiphaseFlow/4comp_2ph_1d_02
compositionalMultiphaseFlow/4comp_2ph_1d_03
compositionalMultiphaseFlow/4comp_2ph_cap_1d_02
compositionalMultiphaseFlow/4comp_2ph_cap_1d_03
compositionalMultiphaseFlow/co2_hybrid_1d_02
compositionalMultiphaseFlow/co2_hybrid_1d_03
compositionalMultiphaseFlow/deadOilSpe10Layers84_85_smoke_2d_04
compositionalMultiphaseFlow/deadoil_3ph_baker_1d_02
compositionalMultiphaseFlow/deadoil_3ph_baker_1d_03
compositionalMultiphaseFlow/deadoil_3ph_corey_1d_02
compositionalMultiphaseFlow/deadoil_3ph_corey_1d_03
compositionalMultiphaseFlow/isothermalLeakyWell_smoke_3d_04
compositionalMultiphaseWell/black_oil_wells_saturated_3d_04
compositionalMultiphaseWell/black_oil_wells_unsaturated_3d_04
compositionalMultiphaseWell/compositional_multiphase_wells_1d_02
compositionalMultiphaseWell/compositional_multiphase_wells_2d_04
compositionalMultiphaseWell/dead_oil_wells_2d_04
compositionalMultiphaseWell/dead_oil_wells_hybrid_2d_04
contactMechanics/ContactMechanics_SingleFracCompression_03
contactMechanics/ContactMechanics_TFrac_04
embeddedFractures/SneddonRotated_smoke_04
embeddedFractures/Sneddon_smoke_04
embeddedFractures/Sneddon_staticCondensation_smoke_04
embeddedFractures/deadoil_3ph_corey_2d_edfm_04
embeddedFractures/deadoil_3ph_corey_2d_impermeableFault_04
embeddedFractures/fractureMatrixFlow_Embedded2d_04
embeddedFractures/fractureMatrixFlow_impermeableEmbFrac_04
embeddedFractures/poroElastic_embFractures_04
embeddedFractures/slipPermeability_pEDFM_04
phaseFieldFracture/PhaseFieldFracture_CohesiveModel_02
phaseFieldFracture/PhaseFieldFracture_SpectralSplit_04
phaseFieldFracture/PhaseFieldFracture_VolDevSplit_04
poroElasticCoupling/PoroDruckerPrager_Wellbore_smoke_2-2-1
poroElasticCoupling/PoroElastic_Mandel_smoke_3-1-1
poroElasticCoupling/PoroElastic_Mandel_smoke_3-1-2
poroElasticCoupling/PoroElastic_Terzaghi_FIM_02
poroElasticCoupling/PoroElastic_Wellbore_smoke_2-2-1
poroElasticCoupling/PoroElastic_deadoil_3ph_baker_2d2-1-2
proppantTransport/FlowProppantBedTransport_2d_02
proppantTransport/FlowProppantBedTransport_2d_03
proppantTransport/FlowProppantTransport_2d_04
proppantTransport/ProppantSlotTest_03
singlePhaseFlow/compressible_1d_02
singlePhaseFlow/compressible_1d_03
singlePhaseFlow/compressible_1d_2solids_02
singlePhaseFlow/compressible_1d_2solids_03
singlePhaseFlow/compressible_2d_2fluids_02
singlePhaseFlow/compressible_2d_2fluids_03
singlePhaseFlow/compressible_2d_2fluids_hybrid_02
singlePhaseFlow/compressible_2d_2fluids_hybrid_03
singlePhaseFlow/fractureFlow_2d_02
singlePhaseFlow/fractureFlow_2d_04
singlePhaseFlow/fractureMatrixFlow_2d_04
singlePhaseFlow/incompressible_1d_02
singlePhaseFlow/incompressible_1d_03
singlePhaseFlow/sourceFlux_1d_02
singlePhaseFlow/sourceFlux_1d_03
singlePhaseFlow/sourceFlux_2d_04
singlePhaseWell/compressible_single_phase_wells_1d_02
singlePhaseWell/compressible_single_phase_wells_hybrid_1d_02
singlePhaseWell/incompressible_single_phase_wells_2d_04
singlePhaseWell/incompressible_single_phase_wells_hybrid_2d_04
solidMechanicsSSLE/gravity_02
wellboreMesh/DeviatedElastic_Wellbore_03
wellboreMesh/DeviatedPoroElasticWellbore_Drilling_04
wellboreMesh/DeviatedPoroElasticWellbore_Injection_04
wellboreMesh/ExtendedDruckerPrager_Wellbore_04
wellboreMesh/KirschProblem_04
wellboreMesh/ModifiedCamClay_Wellbore_04
```

> So, I am fairly sure that this won't impact the results/validity of that integrated test (or any other test)

What makes you feel so confident? In all the above cases, there may be fields that could be useful (other than `nodeList` or `fracturedCells`).

**Comment by CusiniM** [comment]
> 

sorry, maybe my answer was not very clear. I would say that those fields are "correctly" ignored by our packByIndex function because I would not expect them to be packed by index. I am not too sure why we actually try to pack them by index though. I guess that correcting that behavior instead of ignoring them would be better.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 1]
```diff
@@ -229,6 +229,8 @@ void SolidMechanicsLagrangianFEM::registerDataOnMesh( Group & meshBodies )
         setPlotLevel( PlotLevel::NOPLOT ).
         setRestartFlags( RestartFlags::NO_WRITE );
 
+      subRegion.excludeWrappersFromPacking( { viewKeyStruct::elemsAttachedToSendOrReceiveNodesString(),
+                                              viewKeyStruct::elemsNotAttachedToSendOrReceiveNodesString() } );
```
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:233]: What I can say for sure is that they are currently not exchanged 😝

File: src/coreComponents/mesh/EmbeddedSurfaceSubRegion.cpp [hunk 2]
```diff
@@ -73,6 +73,12 @@ EmbeddedSurfaceSubRegion::EmbeddedSurfaceSubRegion( string const & name,
   registerWrapper( viewKeyStruct::surfaceElementToParentPlaneString(), &m_parentPlaneName ).
     setDescription( "A map of surface element to the parent fracture name" );
 
+  excludeWrappersFromPacking( { viewKeyStruct::nodeListString(),
+                                viewKeyStruct::edgeListString(),
+                                viewKeyStruct::surfaceElementsToCellRegionsString(),
+                                viewKeyStruct::surfaceElementsToCellSubRegionsString(),
+                                viewKeyStruct::surfaceElementsToCellIndexString() } );
+
```
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/mesh/EmbeddedSurfaceSubRegion.cpp]: Done.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrea-franceschini) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:233]: I can say that in any Lagrangian contact solver each loop is based on fracture elements. I need the nodes of a fracture element, its connections with other elements (both within the fracture and in the surrounding 3D), but I never need elements from nodes.

**Pull request merged**